<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to fastEmu • fastEmu</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to fastEmu">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastEmu</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_to_fastEmu.html">Introduction to fastEmu</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to fastEmu</h1>
                        <h4 data-toc-skip class="author">Sarah
Teichman</h4>
            
            <h4 data-toc-skip class="date">2025-07-10</h4>
      

      <div class="hidden name"><code>intro_to_fastEmu.Rmd</code></div>

    </div>

    
    
<p>First, we will install <code>radEmu</code> and <code>fastEmu</code>
if we haven’t already.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if (!require("remotes", quietly = TRUE))</span></span>
<span><span class="co">#     install.packages("remotes")</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># remotes::install_github("statdivlab/radEmu")</span></span>
<span><span class="co"># remotes::install_github("statdivlab/fastEmu")</span></span></code></pre></div>
<p>Next, we can load <code>radEmu</code> and <code>fastEmu</code>, as
well as the <code>tidyverse</code> package suite.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/statdivlab/radEmu" class="external-link">radEmu</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">fastEmu</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides an introduction to using <code>fastEmu</code>,
a companion method to <code>radEmu</code>. We will demonstrate this
method using a <a href="https://www.nature.com/articles/s41591-019-0406-6" class="external-link">dataset
published by Wirbel et al. (2019)</a>. This is a meta-analysis of
case-control studies about colorectal cancer.</p>
<p>Wirbel et al. published two pieces of data we’ll focus on today:</p>
<ul>
<li>metadata giving demographics and other information about
participants</li>
<li>a mOTU (metagenomic OTU) table</li>
</ul>
</div>
<div class="section level2">
<h2 id="loading-and-exploring-data">Loading and exploring data<a class="anchor" aria-label="anchor" href="#loading-and-exploring-data"></a>
</h2>
<p>We’ll start by looking at the metadata.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"wirbel_sample"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wirbel_sample</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 566  14</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">wirbel_sample</span><span class="op">)</span></span>
<span><span class="co">#&gt;                             Sample_ID External_ID Age Gender BMI Country  Study</span></span>
<span><span class="co">#&gt; CCIS00146684ST.4.0 CCIS00146684ST-4-0      FR-726  72      F  25     FRA FR-CRC</span></span>
<span><span class="co">#&gt; CCIS00281083ST.3.0 CCIS00281083ST-3-0      FR-060  53      M  32     FRA FR-CRC</span></span>
<span><span class="co">#&gt; CCIS02124300ST.4.0 CCIS02124300ST-4-0      FR-568  35      M  23     FRA FR-CRC</span></span>
<span><span class="co">#&gt; CCIS02379307ST.4.0 CCIS02379307ST-4-0      FR-828  67      M  28     FRA FR-CRC</span></span>
<span><span class="co">#&gt; CCIS02856720ST.4.0 CCIS02856720ST-4-0      FR-027  74      M  27     FRA FR-CRC</span></span>
<span><span class="co">#&gt; CCIS03473770ST.4.0 CCIS03473770ST-4-0      FR-192  29      M  24     FRA FR-CRC</span></span>
<span><span class="co">#&gt;                    Group Library_Size Age_spline.1 Age_spline.2 BMI_spline.1</span></span>
<span><span class="co">#&gt; CCIS00146684ST.4.0   CTR     35443944  -0.19755428    0.7389621   1.18982420</span></span>
<span><span class="co">#&gt; CCIS00281083ST.3.0   CTR     19307896  -0.08126128   -0.6818534  -1.40679307</span></span>
<span><span class="co">#&gt; CCIS02124300ST.4.0   CTR     42141246  -2.17453529   -0.6818534   0.45476676</span></span>
<span><span class="co">#&gt; CCIS02379307ST.4.0   CRC      4829533   0.67464323   -0.1490476   0.07698823</span></span>
<span><span class="co">#&gt; CCIS02856720ST.4.0   CTR     34294675  -0.54643328    1.0941660   0.44793355</span></span>
<span><span class="co">#&gt; CCIS03473770ST.4.0   CTR     20262319  -2.87229329   -0.6818534   0.95261443</span></span>
<span><span class="co">#&gt;                    BMI_spline.2 Sampling</span></span>
<span><span class="co">#&gt; CCIS00146684ST.4.0   -0.5606919   BEFORE</span></span>
<span><span class="co">#&gt; CCIS00281083ST.3.0    2.0039136   BEFORE</span></span>
<span><span class="co">#&gt; CCIS02124300ST.4.0   -0.6706035   BEFORE</span></span>
<span><span class="co">#&gt; CCIS02379307ST.4.0    0.5384247   BEFORE</span></span>
<span><span class="co">#&gt; CCIS02856720ST.4.0    0.1720525   BEFORE</span></span>
<span><span class="co">#&gt; CCIS03473770ST.4.0   -0.6706035   BEFORE</span></span></code></pre></div>
<p>Let’s look at the individual cohorts included in this analysis.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wirbel_sample</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Study</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>count <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 2</span></span></span>
<span><span class="co">#&gt;   Study  count</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> AT-CRC   109</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CN-CRC   126</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> DE-CRC   120</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> FR-CRC   110</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> US-CRC   101</span></span></code></pre></div>
<p>In this vignette, we will focus on samples from the Australia
cohort.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wirbel_sample</span><span class="op">$</span><span class="va">Group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">wirbel_sample</span><span class="op">$</span><span class="va">Group</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTR"</span>,<span class="st">"CRC"</span><span class="op">)</span><span class="op">)</span> <span class="co"># make group a factor</span></span>
<span><span class="va">wirbel_sample</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Study</span> <span class="op">==</span> <span class="st">"AT-CRC"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>count <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   Group count</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> CTR      63</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CRC      46</span></span>
<span><span class="va">wirbel_sample_at</span> <span class="op">&lt;-</span> <span class="va">wirbel_sample</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Study</span> <span class="op">==</span> <span class="st">"AT-CRC"</span><span class="op">)</span></span></code></pre></div>
<p>In the US cohort, we have
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>46</mn><annotation encoding="application/x-tex">46</annotation></semantics></math>
cases and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>63</mn><annotation encoding="application/x-tex">63</annotation></semantics></math>
controls.</p>
<p>Now let’s load the mOTU table.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"wirbel_otu"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wirbel_otu</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 566 845</span></span>
<span><span class="co"># let's check out a subset</span></span>
<span><span class="va">wirbel_otu</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt;                    Streptococcus anginosus [ref_mOTU_v2_0004]</span></span>
<span><span class="co">#&gt; CCIS00146684ST.4.0                                          0</span></span>
<span><span class="co">#&gt; CCIS00281083ST.3.0                                          0</span></span>
<span><span class="co">#&gt; CCIS02124300ST.4.0                                          2</span></span>
<span><span class="co">#&gt; CCIS02379307ST.4.0                                          0</span></span>
<span><span class="co">#&gt; CCIS02856720ST.4.0                                          1</span></span>
<span><span class="co">#&gt;                    Enterobacteriaceae sp. [ref_mOTU_v2_0036]</span></span>
<span><span class="co">#&gt; CCIS00146684ST.4.0                                         3</span></span>
<span><span class="co">#&gt; CCIS00281083ST.3.0                                         0</span></span>
<span><span class="co">#&gt; CCIS02124300ST.4.0                                         5</span></span>
<span><span class="co">#&gt; CCIS02379307ST.4.0                                         5</span></span>
<span><span class="co">#&gt; CCIS02856720ST.4.0                                       675</span></span>
<span><span class="co">#&gt;                    Citrobacter sp. [ref_mOTU_v2_0076]</span></span>
<span><span class="co">#&gt; CCIS00146684ST.4.0                                  0</span></span>
<span><span class="co">#&gt; CCIS00281083ST.3.0                                  0</span></span>
<span><span class="co">#&gt; CCIS02124300ST.4.0                                  0</span></span>
<span><span class="co">#&gt; CCIS02379307ST.4.0                                  0</span></span>
<span><span class="co">#&gt; CCIS02856720ST.4.0                                  0</span></span></code></pre></div>
<p>We can see that this table has
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>566</mn><annotation encoding="application/x-tex">566</annotation></semantics></math>
samples (just like the metadata) and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>845</mn><annotation encoding="application/x-tex">845</annotation></semantics></math>
mOTUs. Let’s subset these samples to those from the Australian
cohort.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wirbel_otu_at</span> <span class="op">&lt;-</span> <span class="va">wirbel_otu</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">wirbel_otu</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">wirbel_sample_at</span><span class="op">$</span><span class="va">Sample_ID</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wirbel_otu_at</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 109 845</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wirbel_sample_at</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 109  14</span></span></code></pre></div>
<p>Now, we are left with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>109</mn><annotation encoding="application/x-tex">109</annotation></semantics></math>
samples.</p>
<p>The <code>radEmu</code> and <code>fastEmu</code> methods can only
consider taxa that appear in at least one sample in the dataset.
Therefore, we must remove any taxa that don’t appear in the Australian
cohort.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">wirbel_otu_at</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="co"># 20 taxa do not appear </span></span>
<span><span class="co">#&gt; [1] 20</span></span>
<span><span class="va">tax_to_rm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">wirbel_otu_at</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">wirbel_otu_at_filt</span> <span class="op">&lt;-</span> <span class="va">wirbel_otu_at</span><span class="op">[</span>, <span class="op">-</span><span class="va">tax_to_rm</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wirbel_otu_at_filt</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 109 825</span></span></code></pre></div>
<p>Now we are left with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>109</mn><annotation encoding="application/x-tex">109</annotation></semantics></math>
samples and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>825</mn><annotation encoding="application/x-tex">825</annotation></semantics></math>
taxa that appear in these samples.</p>
<p>Finally, we can load taxonomic information. We’ll use this
information later.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">wirbel_taxonomy</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">wirbel_taxonomy</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                     domain     phylum          </span></span>
<span><span class="co">#&gt; Streptococcus anginosus [ref_mOTU_v2_0004]          "Bacteria" "Bacillota"     </span></span>
<span><span class="co">#&gt; Enterobacteriaceae sp. [ref_mOTU_v2_0036]           "Bacteria" "Pseudomonadota"</span></span>
<span><span class="co">#&gt; Citrobacter sp. [ref_mOTU_v2_0076]                  "Bacteria" "Pseudomonadota"</span></span>
<span><span class="co">#&gt; Klebsiella michiganensis/oxytoca [ref_mOTU_v2_0079] "Bacteria" "Pseudomonadota"</span></span>
<span><span class="co">#&gt; Enterococcus faecalis [ref_mOTU_v2_0116]            "Bacteria" "Bacillota"     </span></span>
<span><span class="co">#&gt; Lactobacillus salivarius [ref_mOTU_v2_0125]         "Bacteria" "Bacillota"     </span></span>
<span><span class="co">#&gt;                                                     class                </span></span>
<span><span class="co">#&gt; Streptococcus anginosus [ref_mOTU_v2_0004]          "Bacilli"            </span></span>
<span><span class="co">#&gt; Enterobacteriaceae sp. [ref_mOTU_v2_0036]           "Gammaproteobacteria"</span></span>
<span><span class="co">#&gt; Citrobacter sp. [ref_mOTU_v2_0076]                  "Gammaproteobacteria"</span></span>
<span><span class="co">#&gt; Klebsiella michiganensis/oxytoca [ref_mOTU_v2_0079] "Gammaproteobacteria"</span></span>
<span><span class="co">#&gt; Enterococcus faecalis [ref_mOTU_v2_0116]            "Bacilli"            </span></span>
<span><span class="co">#&gt; Lactobacillus salivarius [ref_mOTU_v2_0125]         "Bacilli"            </span></span>
<span><span class="co">#&gt;                                                     order             </span></span>
<span><span class="co">#&gt; Streptococcus anginosus [ref_mOTU_v2_0004]          "Lactobacillales" </span></span>
<span><span class="co">#&gt; Enterobacteriaceae sp. [ref_mOTU_v2_0036]           "Enterobacterales"</span></span>
<span><span class="co">#&gt; Citrobacter sp. [ref_mOTU_v2_0076]                  "Enterobacterales"</span></span>
<span><span class="co">#&gt; Klebsiella michiganensis/oxytoca [ref_mOTU_v2_0079] "Enterobacterales"</span></span>
<span><span class="co">#&gt; Enterococcus faecalis [ref_mOTU_v2_0116]            "Lactobacillales" </span></span>
<span><span class="co">#&gt; Lactobacillus salivarius [ref_mOTU_v2_0125]         "Lactobacillales" </span></span>
<span><span class="co">#&gt;                                                     family              </span></span>
<span><span class="co">#&gt; Streptococcus anginosus [ref_mOTU_v2_0004]          "Streptococcaceae"  </span></span>
<span><span class="co">#&gt; Enterobacteriaceae sp. [ref_mOTU_v2_0036]           "Enterobacteriaceae"</span></span>
<span><span class="co">#&gt; Citrobacter sp. [ref_mOTU_v2_0076]                  "Enterobacteriaceae"</span></span>
<span><span class="co">#&gt; Klebsiella michiganensis/oxytoca [ref_mOTU_v2_0079] "Enterobacteriaceae"</span></span>
<span><span class="co">#&gt; Enterococcus faecalis [ref_mOTU_v2_0116]            "Enterococcaceae"   </span></span>
<span><span class="co">#&gt; Lactobacillus salivarius [ref_mOTU_v2_0125]         "Lactobacillaceae"  </span></span>
<span><span class="co">#&gt;                                                     genus               </span></span>
<span><span class="co">#&gt; Streptococcus anginosus [ref_mOTU_v2_0004]          "Streptococcus"     </span></span>
<span><span class="co">#&gt; Enterobacteriaceae sp. [ref_mOTU_v2_0036]           "Enterobacteriaceae"</span></span>
<span><span class="co">#&gt; Citrobacter sp. [ref_mOTU_v2_0076]                  "Citrobacter"       </span></span>
<span><span class="co">#&gt; Klebsiella michiganensis/oxytoca [ref_mOTU_v2_0079] "Klebsiella"        </span></span>
<span><span class="co">#&gt; Enterococcus faecalis [ref_mOTU_v2_0116]            "Enterococcus"      </span></span>
<span><span class="co">#&gt; Lactobacillus salivarius [ref_mOTU_v2_0125]         "Lactobacillus"     </span></span>
<span><span class="co">#&gt;                                                     species                       </span></span>
<span><span class="co">#&gt; Streptococcus anginosus [ref_mOTU_v2_0004]          "Streptococcus anginosus"     </span></span>
<span><span class="co">#&gt; Enterobacteriaceae sp. [ref_mOTU_v2_0036]           "Enterobacteriaceae bacterium"</span></span>
<span><span class="co">#&gt; Citrobacter sp. [ref_mOTU_v2_0076]                  "Citrobacter sp."             </span></span>
<span><span class="co">#&gt; Klebsiella michiganensis/oxytoca [ref_mOTU_v2_0079] "Klebsiella michiganensis"    </span></span>
<span><span class="co">#&gt; Enterococcus faecalis [ref_mOTU_v2_0116]            "Enterococcus faecalis"       </span></span>
<span><span class="co">#&gt; Lactobacillus salivarius [ref_mOTU_v2_0125]         "Ligilactobacillus salivarius"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fitting-a-model">Fitting a model<a class="anchor" aria-label="anchor" href="#fitting-a-model"></a>
</h2>
<p>We will now fit a model using <code><a href="https://statdivlab.github.io/radEmu/reference/emuFit.html" class="external-link">emuFit()</a></code> from
<code>radEmu</code>. We are interested in studying the log
fold-differences in mOTU abundances across the <code>Group</code>
covariate, while accounting for <code>Gender</code>. In the code below,
we estimated these log fold-difference parameters. This will take up to
a minute to run, depending on your machine.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">emuMod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://statdivlab.github.io/radEmu/reference/emuFit.html" class="external-link">emuFit</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">wirbel_otu_at_filt</span>, formula <span class="op">=</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Gender</span>,</span>
<span>                 data <span class="op">=</span> <span class="va">wirbel_sample_at</span>, run_score_tests <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                 tolerance <span class="op">=</span> <span class="fl">1e-4</span><span class="op">)</span></span></code></pre></div>
<p>Let’s take a look at our results.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">emuMod</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span></span>
<span><span class="co">#&gt;   covariate                                            category category_num</span></span>
<span><span class="co">#&gt; 1  GroupCRC          Streptococcus anginosus [ref_mOTU_v2_0004]            1</span></span>
<span><span class="co">#&gt; 2  GroupCRC           Enterobacteriaceae sp. [ref_mOTU_v2_0036]            2</span></span>
<span><span class="co">#&gt; 3  GroupCRC                  Citrobacter sp. [ref_mOTU_v2_0076]            3</span></span>
<span><span class="co">#&gt; 4  GroupCRC Klebsiella michiganensis/oxytoca [ref_mOTU_v2_0079]            4</span></span>
<span><span class="co">#&gt; 5  GroupCRC            Enterococcus faecalis [ref_mOTU_v2_0116]            5</span></span>
<span><span class="co">#&gt; 6  GroupCRC         Lactobacillus salivarius [ref_mOTU_v2_0125]            6</span></span>
<span><span class="co">#&gt;     estimate        se       lower     upper score_stat pval</span></span>
<span><span class="co">#&gt; 1 -0.3953228 0.4488537 -1.27505983 0.4844143         NA   NA</span></span>
<span><span class="co">#&gt; 2  1.1194867 0.6092440 -0.07460952 2.3135830         NA   NA</span></span>
<span><span class="co">#&gt; 3  2.2890633 1.0746696  0.18274961 4.3953769         NA   NA</span></span>
<span><span class="co">#&gt; 4  2.7344825 1.0933320  0.59159103 4.8773739         NA   NA</span></span>
<span><span class="co">#&gt; 5  0.8669525 1.0484701 -1.18801117 2.9219161         NA   NA</span></span>
<span><span class="co">#&gt; 6 -1.2006590 0.8778756 -2.92126350 0.5199455         NA   NA</span></span></code></pre></div>
<p>For each covariate and mOTU combination we have an estimate, robust
standard error, and confidence interval. The estimate in the first row
of the results is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><mn>0.395</mn></mrow><annotation encoding="application/x-tex">-0.395</annotation></semantics></math>.
We can interpret this as saying that the expected log fold-difference in
abundance of Streptococcus anginosus [ref_mOTU_v2_0004] in colorectal
cancer cases compared to controls (for participants of the same gender)
is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><mn>0.395</mn></mrow><annotation encoding="application/x-tex">-0.395</annotation></semantics></math>
(this corresponds to a fold-difference of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.67</mn><annotation encoding="application/x-tex">0.67</annotation></semantics></math>),
relative to the typical log fold-difference across all mOTUs between
cases and controls. By typical, we mean that we are comparing each
estimated log fold-difference to the approximate median of estimated log
fold-differences of all mOTUs in the analysis. When we say approximate
median, we refer to a function that is similar to the median but has
nicer mathematical properties.</p>
<p>Now, let’s look at estimates for the <code>Group</code> covariate
across all of our mOTUs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">emuMod_group_tax</span> <span class="op">&lt;-</span> <span class="va">emuMod</span><span class="op">$</span><span class="va">coef</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># add in taxonomy information </span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">covariate</span> <span class="op">==</span> <span class="st">"GroupCRC"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">wirbel_taxonomy</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>              <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>category <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">wirbel_taxonomy</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            by <span class="op">=</span> <span class="st">"category"</span><span class="op">)</span></span>
<span><span class="va">emuMod_group_tax</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">estimate</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>ord <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>, xmin <span class="op">=</span> <span class="va">lower</span>, xmax <span class="op">=</span> <span class="va">upper</span>, color <span class="op">=</span> <span class="va">phylum</span>, y <span class="op">=</span> <span class="va">ord</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="st">"Estimate"</span>, color <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimates and intervals for all mOTUs"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_fastEmu_files/figure-html/unnamed-chunk-13-1.png" width="576"></p>
<p>Here, we can see a range of confidence intervals for estimated log
fold-differences. While the approximate median of the estimates is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>
(this is enforced due to the parameter that we are estimating), the
estimates range from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><mn>7.88</mn></mrow><annotation encoding="application/x-tex">-7.88</annotation></semantics></math>
to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>8.38</mn><annotation encoding="application/x-tex">8.38</annotation></semantics></math>.
Here we can see that the majority of the mOTUs in our analysis belong to
the <em>Bacillota</em> phylum.</p>
<p>Let’s now investigate mOTUs from the <em>Fusobacteriota</em> phylum.
In Wirbel et al.’s analysis, they found some evidence of enrichment for
some members of this phylum. Let’s look at the log fold-difference
estimates for these mOTUs.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">emuMod_group_tax</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">estimate</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">phylum</span> <span class="op">==</span> <span class="st">"Fusobacteriota"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>ord <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>, xmin <span class="op">=</span> <span class="va">lower</span>, xmax <span class="op">=</span> <span class="va">upper</span>, y <span class="op">=</span> <span class="va">ord</span>, color <span class="op">=</span> <span class="va">species</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="st">"Estimate"</span>, color <span class="op">=</span> <span class="st">"Species"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimates and intervals for Fusobacteriota"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_fastEmu_files/figure-html/unnamed-chunk-14-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="running-hypothesis-tests">Running hypothesis tests<a class="anchor" aria-label="anchor" href="#running-hypothesis-tests"></a>
</h2>
<div class="section level3">
<h3 id="testing-with-rademu">Testing with <code>radEmu</code><a class="anchor" aria-label="anchor" href="#testing-with-rademu"></a>
</h3>
<p>Now that we’ve explored our estimates, let’s move on to running
hypothesis tests. Specifically, let’s test the null hypotheses that the
expected log fold-difference in abundance in colorectal cases compared
controls (accounting for gender) for each <em>Fusobacteriota</em> mOTU
is equal to the typical log fold-difference between cases and controls
across all mOTUs.</p>
<p>First, we’ll pull out the indices of these mOTUs that we would like
to test.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fuso</span> <span class="op">&lt;-</span> <span class="va">emuMod_group_tax</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">phylum</span> <span class="op">==</span> <span class="st">"Fusobacteriota"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">category</span><span class="op">)</span></span>
<span><span class="va">fuso_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">wirbel_otu_at_filt</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">fuso</span><span class="op">)</span></span></code></pre></div>
<p><code>radEmu</code> uses robust score tests for hypothesis testing.
These tests are robust to model misspecification and are shown to
control the Type I error rate in a variety of simulation settings,
including those with small sample sizes and datasets with a large
proportion of zeros. However, one downside of these tests is that each
test requires a substantial amount of computational time, and this gets
larger for datasets with more taxa.</p>
<p>Below, we have code to test the nine <em>Fusobacteriota</em> mOTUs.
On a Mac with 16 GB of memory and the M4 chip, this takes approximately
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>6.5</mn><annotation encoding="application/x-tex">6.5</annotation></semantics></math>
minutes to run. You can either run this below, or load in the
results.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">radFusTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://statdivlab.github.io/radEmu/reference/emuFit.html" class="external-link">emuFit</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">wirbel_otu_at_filt</span>, formula <span class="op">=</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Gender</span>,</span>
<span>                     data <span class="op">=</span> <span class="va">wirbel_sample_at</span>, fitted_model <span class="op">=</span> <span class="va">emuMod</span>, </span>
<span>                     refit <span class="op">=</span> <span class="cn">FALSE</span>, compute_cis <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                     test_kj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">2</span>, j <span class="op">=</span> <span class="va">fuso_ind</span><span class="op">)</span>,</span>
<span>                     verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">radFusTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"radFusTest.rds"</span>, package <span class="op">=</span> <span class="st">"fastEmu"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">radTimes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">23</span>, <span class="fl">27</span>, <span class="fl">198</span>, <span class="fl">18</span>, <span class="fl">8</span>, <span class="fl">26</span>, <span class="fl">15</span>, <span class="fl">15</span>, <span class="fl">56</span><span class="op">)</span> <span class="co"># the times each score test took on my laptop</span></span></code></pre></div>
<p>While
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>6.5</mn><annotation encoding="application/x-tex">6.5</annotation></semantics></math>
minutes isn’t very long to wait, you can imagine that testing all
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>825</mn><annotation encoding="application/x-tex">825</annotation></semantics></math>
mOTUs would take a long time, and if we had a larger set of mOTUs
(several thousand for example), each test would take longer.</p>
</div>
<div class="section level3">
<h3 id="testing-with-fastemu">Testing with <code>fastEmu</code><a class="anchor" aria-label="anchor" href="#testing-with-fastemu"></a>
</h3>
<p><code>fastEmu</code> is a companion method to <code>radEmu</code>
that runs robust score tests faster for approximately the same
parameters that we estimate and test with <code>radEmu</code>. This is
especially useful for large datasets, if using <code>radEmu</code> is
computationally infeasible given your computational resources.</p>
<p>We will start by comparing the parameters estimated by
<code>fastEmu</code> to <code>radEmu</code>. <code>fastEmu</code> works
by fitting a simpler model than the one used by <code>radEmu</code>,
while targeting a very similar parameter. In order to define this
simpler model, we need a parameter that relies on a subset of taxa, as
opposed to one defined relative to the “typical log fold-difference
across all taxa.” In <code>fastEmu</code>, we instead compare to the
“typical log fold-difference across taxa in a reference set.” Ideally,
the approximate median log fold-difference in this reference set would
be quite similar to the approximate median log fold-difference in the
full analysis.</p>
<p>In <code>fastEmu</code>, the user can either decide which taxa to
include in this reference set, or a “data-driven” reference set can be
determined. This “data-driven” reference set is a small set of taxa that
have estimated log fold-differences that are closest to the approximate
median of estimated log fold-differences from <code>radEmu</code>.
Below, we can use <code><a href="../reference/fastEmuFit.html">fastEmuFit()</a></code> to determine a “data-driven”
reference set.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastMod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastEmuFit.html">fastEmuFit</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">wirbel_otu_at_filt</span>, formula <span class="op">=</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Gender</span>,</span>
<span>                      data <span class="op">=</span> <span class="va">wirbel_sample_at</span>, fitted_model <span class="op">=</span> <span class="va">emuMod</span>, </span>
<span>                      reference_set <span class="op">=</span> <span class="st">"data_driven"</span>, run_score_tests <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">fastMod</span><span class="op">$</span><span class="va">reference_set_names</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;  [1] "unknown Roseburia [meta_mOTU_v2_7567]"              </span></span>
<span><span class="co">#&gt;  [2] "unknown Clostridiales [meta_mOTU_v2_6575]"          </span></span>
<span><span class="co">#&gt;  [3] "unknown Clostridiales [meta_mOTU_v2_7031]"          </span></span>
<span><span class="co">#&gt;  [4] "unknown Lactobacillales [meta_mOTU_v2_6288]"        </span></span>
<span><span class="co">#&gt;  [5] "unknown Clostridiales [meta_mOTU_v2_5954]"          </span></span>
<span><span class="co">#&gt;  [6] "unknown Clostridiales [meta_mOTU_v2_6134]"          </span></span>
<span><span class="co">#&gt;  [7] "unknown Clostridiales [meta_mOTU_v2_6751]"          </span></span>
<span><span class="co">#&gt;  [8] "Eubacterium sp. CAG:156 [meta_mOTU_v2_7325]"        </span></span>
<span><span class="co">#&gt;  [9] "unknown Ruminococcaceae [meta_mOTU_v2_7271]"        </span></span>
<span><span class="co">#&gt; [10] "unknown Clostridium [meta_mOTU_v2_6816]"            </span></span>
<span><span class="co">#&gt; [11] "Niameybacter massiliensis [meta_mOTU_v2_7610]"      </span></span>
<span><span class="co">#&gt; [12] "unknown Azospirillum [meta_mOTU_v2_6527]"           </span></span>
<span><span class="co">#&gt; [13] "unknown Clostridiales [meta_mOTU_v2_7707]"          </span></span>
<span><span class="co">#&gt; [14] "unknown Desulfovibrio [meta_mOTU_v2_5993]"          </span></span>
<span><span class="co">#&gt; [15] "unknown Clostridiales [meta_mOTU_v2_7087]"          </span></span>
<span><span class="co">#&gt; [16] "unknown Clostridiales [meta_mOTU_v2_7781]"          </span></span>
<span><span class="co">#&gt; [17] "Desulfovibrio piger [ref_mOTU_v2_4205]"             </span></span>
<span><span class="co">#&gt; [18] "unknown Clostridiales [meta_mOTU_v2_6856]"          </span></span>
<span><span class="co">#&gt; [19] "Bifidobacterium ruminantium [ref_mOTU_v2_1159]"     </span></span>
<span><span class="co">#&gt; [20] "Intestinimonas butyriciproducens [ref_mOTU_v2_2968]"</span></span>
<span><span class="co">#&gt; [21] "unknown Clostridiales [meta_mOTU_v2_7415]"          </span></span>
<span><span class="co">#&gt; [22] "Clostridium sp. CAG:1193 [meta_mOTU_v2_7613]"       </span></span>
<span><span class="co">#&gt; [23] "unknown Clostridiales [meta_mOTU_v2_7800]"          </span></span>
<span><span class="co">#&gt; [24] "unknown Firmicutes [meta_mOTU_v2_6277]"             </span></span>
<span><span class="co">#&gt; [25] "unknown Bacteroidales [meta_mOTU_v2_5329]"          </span></span>
<span><span class="co">#&gt; [26] "unknown Clostridiales [meta_mOTU_v2_6416]"          </span></span>
<span><span class="co">#&gt; [27] "unknown Acetobacter [meta_mOTU_v2_6063]"            </span></span>
<span><span class="co">#&gt; [28] "unknown Bacteroidales [meta_mOTU_v2_6591]"          </span></span>
<span><span class="co">#&gt; [29] "unknown Clostridiales [meta_mOTU_v2_5991]"          </span></span>
<span><span class="co">#&gt; [30] "unknown Atopobiaceae [meta_mOTU_v2_6458]"           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;  [1] "unknown Firmicutes [meta_mOTU_v2_7689]"             </span></span>
<span><span class="co">#&gt;  [2] "unknown Clostridiales [meta_mOTU_v2_5791]"          </span></span>
<span><span class="co">#&gt;  [3] "Holdemania filiformis [ref_mOTU_v2_4459]"           </span></span>
<span><span class="co">#&gt;  [4] "unknown Clostridiales [meta_mOTU_v2_5347]"          </span></span>
<span><span class="co">#&gt;  [5] "unknown Clostridiales [meta_mOTU_v2_6885]"          </span></span>
<span><span class="co">#&gt;  [6] "unknown Clostridiales [meta_mOTU_v2_5653]"          </span></span>
<span><span class="co">#&gt;  [7] "Eubacterium sp. CAG:274 [meta_mOTU_v2_7140]"        </span></span>
<span><span class="co">#&gt;  [8] "unknown Clostridiales [meta_mOTU_v2_7093]"          </span></span>
<span><span class="co">#&gt;  [9] "Enterococcus faecium [ref_mOTU_v2_0372]"            </span></span>
<span><span class="co">#&gt; [10] "unknown Bacteroidales [meta_mOTU_v2_5951]"          </span></span>
<span><span class="co">#&gt; [11] "Clostridium spiroforme [ref_mOTU_v2_4235]"          </span></span>
<span><span class="co">#&gt; [12] "Desulfovibrio piger [ref_mOTU_v2_4205]"             </span></span>
<span><span class="co">#&gt; [13] "Clostridium sp. 27_14 [meta_mOTU_v2_5718]"          </span></span>
<span><span class="co">#&gt; [14] "Anaeromassilibacillus sp. An200 [meta_mOTU_v2_6051]"</span></span>
<span><span class="co">#&gt; [15] "Alistipes shahii [ref_mOTU_v2_4873]"                </span></span>
<span><span class="co">#&gt; [16] "unknown Faecalibacterium [meta_mOTU_v2_5815]"       </span></span>
<span><span class="co">#&gt; [17] "unknown Clostridiales [meta_mOTU_v2_7014]"          </span></span>
<span><span class="co">#&gt; [18] "Clostridium sp. CAG:343 [meta_mOTU_v2_6218]"        </span></span>
<span><span class="co">#&gt; [19] "unknown Clostridiales [meta_mOTU_v2_7180]"          </span></span>
<span><span class="co">#&gt; [20] "unknown Clostridiales [meta_mOTU_v2_6808]"          </span></span>
<span><span class="co">#&gt; [21] "Clostridium sp. CAG:288 [meta_mOTU_v2_6160]"        </span></span>
<span><span class="co">#&gt; [22] "unknown Clostridiales [meta_mOTU_v2_6801]"          </span></span>
<span><span class="co">#&gt; [23] "Megamonas funiformis/rupellensis [ref_mOTU_v2_0502]"</span></span>
<span><span class="co">#&gt; [24] "unknown Clostridiales [meta_mOTU_v2_6105]"          </span></span>
<span><span class="co">#&gt; [25] "unknown Clostridiales [meta_mOTU_v2_5339]"          </span></span>
<span><span class="co">#&gt; [26] "unknown Peptostreptococcaceae [meta_mOTU_v2_7331]"  </span></span>
<span><span class="co">#&gt; [27] "Clostridium sp. JCC [ref_mOTU_v2_3353]"             </span></span>
<span><span class="co">#&gt; [28] "unknown Clostridiales [meta_mOTU_v2_5805]"          </span></span>
<span><span class="co">#&gt; [29] "unknown Roseburia [meta_mOTU_v2_5354]"              </span></span>
<span><span class="co">#&gt; [30] "Ruminococcus lactaris [ref_mOTU_v2_0281]"           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt;  [1] "Porphyromonas uenonis [ref_mOTU_v2_4616]"              </span></span>
<span><span class="co">#&gt;  [2] "Ruminococcus bicirculans [ref_mOTU_v2_2358]"           </span></span>
<span><span class="co">#&gt;  [3] "Ruminococcus gnavus [ref_mOTU_v2_0280]"                </span></span>
<span><span class="co">#&gt;  [4] "Eubacterium eligens [ref_mOTU_v2_4389]"                </span></span>
<span><span class="co">#&gt;  [5] "Atopobium parvulum [ref_mOTU_v2_0741]"                 </span></span>
<span><span class="co">#&gt;  [6] "unknown Clostridiales [meta_mOTU_v2_7769]"             </span></span>
<span><span class="co">#&gt;  [7] "unknown Anaerotruncus [meta_mOTU_v2_6835]"             </span></span>
<span><span class="co">#&gt;  [8] "Streptococcus anginosus [ref_mOTU_v2_0004]"            </span></span>
<span><span class="co">#&gt;  [9] "unknown Firmicutes [meta_mOTU_v2_6711]"                </span></span>
<span><span class="co">#&gt; [10] "unknown Clostridiales [meta_mOTU_v2_7337]"             </span></span>
<span><span class="co">#&gt; [11] "Alistipes senegalensis [ref_mOTU_v2_1594]"             </span></span>
<span><span class="co">#&gt; [12] "unknown Sutterella [meta_mOTU_v2_5929]"                </span></span>
<span><span class="co">#&gt; [13] "unknown Clostridiales [meta_mOTU_v2_5514]"             </span></span>
<span><span class="co">#&gt; [14] "unknown Clostridiales [meta_mOTU_v2_7186]"             </span></span>
<span><span class="co">#&gt; [15] "Coprobacillus sp. CAG:605 [meta_mOTU_v2_7625]"         </span></span>
<span><span class="co">#&gt; [16] "Bilophila wadsworthia [ref_mOTU_v2_1149]"              </span></span>
<span><span class="co">#&gt; [17] "unknown Clostridiales [meta_mOTU_v2_7782]"             </span></span>
<span><span class="co">#&gt; [18] "Eubacterium ramulus [ref_mOTU_v2_2795]"                </span></span>
<span><span class="co">#&gt; [19] "unknown Clostridiales [meta_mOTU_v2_7455]"             </span></span>
<span><span class="co">#&gt; [20] "unknown Ruminococcaceae [meta_mOTU_v2_6557]"           </span></span>
<span><span class="co">#&gt; [21] "unknown Eubacterium [meta_mOTU_v2_6509]"               </span></span>
<span><span class="co">#&gt; [22] "unknown Clostridiales [meta_mOTU_v2_6852]"             </span></span>
<span><span class="co">#&gt; [23] "unknown Lactobacillales [meta_mOTU_v2_6288]"           </span></span>
<span><span class="co">#&gt; [24] "Fusobacterium nucleatum s. animalis [ref_mOTU_v2_0776]"</span></span>
<span><span class="co">#&gt; [25] "uncultured Eubacterium sp. [meta_mOTU_v2_5477]"        </span></span>
<span><span class="co">#&gt; [26] "Granulicatella adiacens [ref_mOTU_v2_4659]"            </span></span>
<span><span class="co">#&gt; [27] "Turicibacter sanguinis [ref_mOTU_v2_1493]"             </span></span>
<span><span class="co">#&gt; [28] "Clostridium sp. CAG:127 [meta_mOTU_v2_5364]"           </span></span>
<span><span class="co">#&gt; [29] "unknown Clostridiales [meta_mOTU_v2_7527]"             </span></span>
<span><span class="co">#&gt; [30] "unknown Clostridiales [meta_mOTU_v2_7553]"</span></span>
<span><span class="va">fastMod</span><span class="op">$</span><span class="va">constraint_diff</span></span>
<span><span class="co">#&gt; [1] -0.006124724  0.006670540  0.001169057</span></span></code></pre></div>
<p>Above, we can see the reference sets chosen for each covariate in the
model (as well as the intercept). The <code>constraint_diff</code>
object tells us the difference between the approximate median log
fold-difference across all taxa and the approximate median log
fold-difference across the reference set, for each covariate. Each of
these differences represents a small shift in all parameter estimates
from <code>radEmu</code> to <code>fastEmu</code>.</p>
<p>Now, we can run our robust score tests for the
<em>Fusobacteriota</em> mOTUs using <code>fastEmu</code>. We’ll pass in
our fit above <code>fastMod</code> in order to use the “data-driven”
reference set that we already computed.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastFusTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastEmuFit.html">fastEmuFit</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">wirbel_otu_at_filt</span>, formula <span class="op">=</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Gender</span>,</span>
<span>                          data <span class="op">=</span> <span class="va">wirbel_sample_at</span>, fitted_model <span class="op">=</span> <span class="va">fastMod</span>, </span>
<span>                          test_kj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">2</span>, j <span class="op">=</span> <span class="va">fuso_ind</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Skipping estimation and reference set building/checking because a `fastEmuFit` fitted model has been provided. Proceeding immediately with score tests.</span></span>
<span><span class="co">#&gt; Running score test 1 of 9.</span></span>
<span><span class="co">#&gt; Score test 1 of 9 has completed in approximately 5 seconds.</span></span>
<span><span class="co">#&gt; Running score test 2 of 9.</span></span>
<span><span class="co">#&gt; Score test 2 of 9 has completed in approximately 6 seconds.</span></span>
<span><span class="co">#&gt; Running score test 3 of 9.</span></span>
<span><span class="co">#&gt; Score test 3 of 9 has completed in approximately 50 seconds.</span></span>
<span><span class="co">#&gt; Running score test 4 of 9.</span></span>
<span><span class="co">#&gt; Score test 4 of 9 has completed in approximately 30 seconds.</span></span>
<span><span class="co">#&gt; Running score test 5 of 9.</span></span>
<span><span class="co">#&gt; Score test 5 of 9 has completed in approximately 3 seconds.</span></span>
<span><span class="co">#&gt; Running score test 6 of 9.</span></span>
<span><span class="co">#&gt; Score test 6 of 9 has completed in approximately 11 seconds.</span></span>
<span><span class="co">#&gt; Running score test 7 of 9.</span></span>
<span><span class="co">#&gt; Score test 7 of 9 has completed in approximately 13 seconds.</span></span>
<span><span class="co">#&gt; Running score test 8 of 9.</span></span>
<span><span class="co">#&gt; Score test 8 of 9 has completed in approximately 4 seconds.</span></span>
<span><span class="co">#&gt; Running score test 9 of 9.</span></span>
<span><span class="co">#&gt; Score test 9 of 9 has completed in approximately 14 seconds.</span></span>
<span><span class="va">fastFusTest</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="va">fuso_ind</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;     covariate                                                category</span></span>
<span><span class="co">#&gt; 64   GroupCRC Fusobacterium nucleatum s. vincentii [ref_mOTU_v2_0754]</span></span>
<span><span class="co">#&gt; 67   GroupCRC  Fusobacterium nucleatum s. animalis [ref_mOTU_v2_0776]</span></span>
<span><span class="co">#&gt; 68   GroupCRC Fusobacterium nucleatum s. nucleatum [ref_mOTU_v2_0777]</span></span>
<span><span class="co">#&gt; 119  GroupCRC               Fusobacterium ulcerans [ref_mOTU_v2_1396]</span></span>
<span><span class="co">#&gt; 122  GroupCRC     Fusobacterium sp. oral taxon 370 [ref_mOTU_v2_1403]</span></span>
<span><span class="co">#&gt; 123  GroupCRC         Fusobacterium gonidiaformans [ref_mOTU_v2_1404]</span></span>
<span><span class="co">#&gt; 185  GroupCRC             Fusobacterium mortiferum [ref_mOTU_v2_4310]</span></span>
<span><span class="co">#&gt; 186  GroupCRC                 Fusobacterium varium [ref_mOTU_v2_4311]</span></span>
<span><span class="co">#&gt; 727  GroupCRC               unknown Fusobacterium [meta_mOTU_v2_7372]</span></span>
<span><span class="co">#&gt;     category_num  estimate        se      lower    upper score_stat        pval</span></span>
<span><span class="co">#&gt; 64            64 1.6223985 0.7229090  0.2055229 3.039274  2.5419415 0.110858395</span></span>
<span><span class="co">#&gt; 67            67 3.2766977 0.5738113  2.1520481 4.401347  7.7815346 0.005278294</span></span>
<span><span class="co">#&gt; 68            68 3.2134419 0.5473542  2.1406475 4.286236  3.0827028 0.079128927</span></span>
<span><span class="co">#&gt; 119          119 1.1324033 0.7549393 -0.3472506 2.612057  1.7469445 0.186261316</span></span>
<span><span class="co">#&gt; 122          122 6.4280384 0.9647693  4.5371254 8.318951  1.1571196 0.282063704</span></span>
<span><span class="co">#&gt; 123          123 4.4786902 0.6838872  3.1382960 5.819085  2.1574059 0.141884047</span></span>
<span><span class="co">#&gt; 185          185 0.5747373 0.6236383 -0.6475713 1.797046  0.8278341 0.362899558</span></span>
<span><span class="co">#&gt; 186          186 3.4603537 0.8871824  1.7215082 5.199199  1.2330106 0.266822340</span></span>
<span><span class="co">#&gt; 727          727 2.4687183 0.6246063  1.2445124 3.692924  2.1750993 0.140260466</span></span>
<span><span class="co">#&gt;     reference_category</span></span>
<span><span class="co">#&gt; 64               FALSE</span></span>
<span><span class="co">#&gt; 67               FALSE</span></span>
<span><span class="co">#&gt; 68               FALSE</span></span>
<span><span class="co">#&gt; 119              FALSE</span></span>
<span><span class="co">#&gt; 122              FALSE</span></span>
<span><span class="co">#&gt; 123              FALSE</span></span>
<span><span class="co">#&gt; 185              FALSE</span></span>
<span><span class="co">#&gt; 186              FALSE</span></span>
<span><span class="co">#&gt; 727              FALSE</span></span>
<span><span class="va">fastTimes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">20</span>, <span class="fl">13</span>, <span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">1</span>, <span class="fl">6</span><span class="op">)</span> <span class="co"># the times each score test took on my laptop</span></span></code></pre></div>
<p>One note is that you don’t need to run <code>radEmu</code> before
<code>fastEmu</code>. Instead you could run the below code, which will
estimate <code>radEmu</code> parameters, generate the “data-drive”
reference set, shift estimates, and then run the desired robust score
tests.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastFusTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastEmuFit.html">fastEmuFit</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">wirbel_otu_at_filt</span>, formula <span class="op">=</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Gender</span>,</span>
<span>                          data <span class="op">=</span> <span class="va">wirbel_sample_at</span>, reference_set <span class="op">=</span> <span class="st">"data_driven"</span>,</span>
<span>                          test_kj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">2</span>, j <span class="op">=</span> <span class="va">fuso_ind</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now, we can compare the test results between <code>radEmu</code> and
<code>fastEmu</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>rad <span class="op">=</span> <span class="va">radFusTest</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">pval</span><span class="op">[</span><span class="va">fuso_ind</span><span class="op">]</span>,</span>
<span>           fast <span class="op">=</span> <span class="va">fastFusTest</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">pval</span><span class="op">[</span><span class="va">fuso_ind</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">rad</span>, y <span class="op">=</span> <span class="va">fast</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"radEmu p-value"</span>, y <span class="op">=</span> <span class="st">"fastEmu p-value"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_to_fastEmu_files/figure-html/unnamed-chunk-21-1.png" width="576"></p>
<p>In this plot, we can see that the p-values from <code>fastEmu</code>
and <code>radEmu</code> for these nine tests are nearly identical.</p>
<p>We can also compare the times that these took (these are the times
from my laptop, you may find the times to run each method to be slightly
different).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">radTimes</span> <span class="op">/</span> <span class="va">fastTimes</span></span>
<span><span class="co">#&gt; [1]  7.666667 13.500000  9.900000  1.384615  8.000000  5.200000  3.000000</span></span>
<span><span class="co">#&gt; [8] 15.000000  9.333333</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">radTimes</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">fastTimes</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6.892857</span></span></code></pre></div>
<p>Overall, running <code>fastEmu</code> was approximately
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>7</mn><annotation encoding="application/x-tex">7</annotation></semantics></math>
times faster than running <code>radEmu</code>. For individual score
tests, <code>fastEmu</code> was between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1.4</mn><annotation encoding="application/x-tex">1.4</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>15</mn><annotation encoding="application/x-tex">15</annotation></semantics></math>
times faster.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastFusTest</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="va">fuso_ind</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;     covariate                                                category</span></span>
<span><span class="co">#&gt; 64   GroupCRC Fusobacterium nucleatum s. vincentii [ref_mOTU_v2_0754]</span></span>
<span><span class="co">#&gt; 67   GroupCRC  Fusobacterium nucleatum s. animalis [ref_mOTU_v2_0776]</span></span>
<span><span class="co">#&gt; 68   GroupCRC Fusobacterium nucleatum s. nucleatum [ref_mOTU_v2_0777]</span></span>
<span><span class="co">#&gt; 119  GroupCRC               Fusobacterium ulcerans [ref_mOTU_v2_1396]</span></span>
<span><span class="co">#&gt; 122  GroupCRC     Fusobacterium sp. oral taxon 370 [ref_mOTU_v2_1403]</span></span>
<span><span class="co">#&gt; 123  GroupCRC         Fusobacterium gonidiaformans [ref_mOTU_v2_1404]</span></span>
<span><span class="co">#&gt; 185  GroupCRC             Fusobacterium mortiferum [ref_mOTU_v2_4310]</span></span>
<span><span class="co">#&gt; 186  GroupCRC                 Fusobacterium varium [ref_mOTU_v2_4311]</span></span>
<span><span class="co">#&gt; 727  GroupCRC               unknown Fusobacterium [meta_mOTU_v2_7372]</span></span>
<span><span class="co">#&gt;     category_num  estimate        se      lower    upper score_stat        pval</span></span>
<span><span class="co">#&gt; 64            64 1.6223985 0.7229090  0.2055229 3.039274  2.5419415 0.110858395</span></span>
<span><span class="co">#&gt; 67            67 3.2766977 0.5738113  2.1520481 4.401347  7.7815346 0.005278294</span></span>
<span><span class="co">#&gt; 68            68 3.2134419 0.5473542  2.1406475 4.286236  3.0827028 0.079128927</span></span>
<span><span class="co">#&gt; 119          119 1.1324033 0.7549393 -0.3472506 2.612057  1.7469445 0.186261316</span></span>
<span><span class="co">#&gt; 122          122 6.4280384 0.9647693  4.5371254 8.318951  1.1571196 0.282063704</span></span>
<span><span class="co">#&gt; 123          123 4.4786902 0.6838872  3.1382960 5.819085  2.1574059 0.141884047</span></span>
<span><span class="co">#&gt; 185          185 0.5747373 0.6236383 -0.6475713 1.797046  0.8278341 0.362899558</span></span>
<span><span class="co">#&gt; 186          186 3.4603537 0.8871824  1.7215082 5.199199  1.2330106 0.266822340</span></span>
<span><span class="co">#&gt; 727          727 2.4687183 0.6246063  1.2445124 3.692924  2.1750993 0.140260466</span></span>
<span><span class="co">#&gt;     reference_category</span></span>
<span><span class="co">#&gt; 64               FALSE</span></span>
<span><span class="co">#&gt; 67               FALSE</span></span>
<span><span class="co">#&gt; 68               FALSE</span></span>
<span><span class="co">#&gt; 119              FALSE</span></span>
<span><span class="co">#&gt; 122              FALSE</span></span>
<span><span class="co">#&gt; 123              FALSE</span></span>
<span><span class="co">#&gt; 185              FALSE</span></span>
<span><span class="co">#&gt; 186              FALSE</span></span>
<span><span class="co">#&gt; 727              FALSE</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="interpreting-results">Interpreting results<a class="anchor" aria-label="anchor" href="#interpreting-results"></a>
</h2>
<p>When we look at the results, we can see that the p-values from the
robust score tests range from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.005</mn><annotation encoding="application/x-tex">0.005</annotation></semantics></math>
to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.363</mn><annotation encoding="application/x-tex">0.363</annotation></semantics></math>.
There is only one taxon with a p-value below
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.05</mn><annotation encoding="application/x-tex">0.05</annotation></semantics></math>,
which is Fusobacterium nucleatum s. animalis [ref_mOTU_v2_0776].</p>
<p>We can account for multiple testing by adjusting our p-values to
control the false discovery rate using the Benjamini-Hochberg procedure
(although if you have a larger number of tests, we recommend using the
qvalue procedure implemented in the <code>qvalue</code> package).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">fastFusTest</span><span class="op">$</span><span class="va">coef</span><span class="op">$</span><span class="va">pval</span><span class="op">[</span><span class="va">fuso_ind</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.25539128 0.04750465 0.25539128 0.27939197 0.31732167 0.25539128 0.36289956</span></span>
<span><span class="co">#&gt; [8] 0.31732167 0.25539128</span></span></code></pre></div>
<p>Here, we can see that at a 5% false discovery rate threshold, we can
reject the null hypothesis that the expected log fold-difference in the
abundance of Fusobacterium nucleatum s. animalis [ref_mOTU_v2_0776] in
colorectal cases compared to controls is equal to the typical log
fold-difference across all taxa between cases and controls.</p>
</div>
<div class="section level2">
<h2 id="final-notes">Final notes<a class="anchor" aria-label="anchor" href="#final-notes"></a>
</h2>
<p>In this analysis, using <code>fastEmu</code> instead of
<code>radEmu</code> only saved us a few minutes. However, in a large
analysis, this may add up. Additionally, the time that
<code>fastEmu</code> saves will get larger and larger for analyses with
more and more taxa. The ideal setting to use <code>fastEmu</code> is
when <code>J</code> (the number of taxa) is quite large and
<code>p</code> (the number of covariates) is small to moderate.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Sarah Teichman, Amy D Willis.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
